numpy/core/src/multiarray/item_selection.c

```diff
@@ -2202,7 +2206,173 @@ count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
     return unrollx - zero_count;
 }
 
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_uintp
+count_nonzero_int16_simd(npy_int16 *d, npy_uintp unrollx)
+{
+    npy_uintp zero_count = 0;
+    const npy_uintp innerloop_jump = NPY_MAX_UINT16;
+    const npy_int16 *end = d + unrollx;
+
+    const npyv_u16 vone = npyv_setall_u16(1); 
+    const npyv_u16 vzero = npyv_zero_u16();   
+
+    npy_int16 *target = d;
+    npy_uint16 sums[npyv_nlanes_u16];
+
+    while (d<end) {
+        npyv_u16 vsum16 = npyv_zero_u16(); 
+        target = PyArray_MIN(target+innerloop_jump, end);
+        for (; d<target; d+=npyv_nlanes_u16) {
+            npyv_u16 vt = npyv_cvt_u16_b16(npyv_cmpeq_u16(npyv_load_u16(d), vzero)); 
+            vt = npyv_and_u16(vt, vone);
+            vsum16 = npyv_add_u16(vsum16, vt);
+        }
+
+        npyv_store_u16(sums, vsum16);
+        for (int i=0; i<npyv_nlanes_u16; ++i) {
+            zero_count += sums[i];
+        }
+    }
+
+    return unrollx - zero_count;
+}
+
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_uintp
+count_nonzero_int32_simd(npy_int32 *d, npy_uintp unrollx)
+{
+    npy_uintp zero_count = 0;
+    const npy_uintp innerloop_jump = NPY_MAX_UINT32;
+    const npy_int32 *end = d + unrollx;
+
+    const npyv_u32 vone = npyv_setall_u32(1); 
+    const npyv_u32 vzero = npyv_zero_u32();   
+
+    npy_int32 *target = d;
+    while (d<end) {
+        npyv_u32 vsum32 = npyv_zero_u32(); 
+        target = PyArray_MIN(target+innerloop_jump, end);
+        for (; d<target; d+=npyv_nlanes_u32) {
+            npyv_u32 vt = npyv_cvt_u32_b32(npyv_cmpeq_u32(npyv_load_u32(d), vzero)); 
+            vt = npyv_and_u32(vt, vone);
+            vsum32 = npyv_add_u32(vsum32, vt);
+        }
+        zero_count += npyv_sum_u32(vsum32);    
+    }
+
+    return unrollx - zero_count;
+}
+
+
+static NPY_INLINE NPY_GCC_OPT_3 npy_uintp
+count_nonzero_int64_simd(npy_int64 *d, npy_uintp unrollx)
+{
+    npy_uintp zero_count = 0;
+    const npy_int64 *end = d + unrollx;
+    const npyv_u64 vone = npyv_setall_u64(1); 
+    const npyv_u64 vzero = npyv_zero_u64();   
+    npyv_u64 vsum64 = npyv_zero_u64(); 
+
+    for (; d<end; d+=npyv_nlanes_u64) {
+        npyv_u64 vt = npyv_cvt_u64_b64(npyv_cmpeq_u64(npyv_load_u64(d), vzero)); 
+        vt = npyv_and_u64(vt, vone);
+        vsum64 = npyv_add_u64(vsum64, vt);
+    }
+
+    npy_uint64 sums[npyv_nlanes_u64];
```

Comment:
you can use new acceleration intrinsics after #18200 merged.

Reply:
@Qiyu8 , I have replaced the manual sums with horizontal SIMD sums.

Reply:
Well done, The replaced part looks good to me, Now you need to focus on fix the CI failures and provide ASV benchmark result.